Steam - second project pull request

DvdHalliday opened the pull request:
Added reference-friendly behaviour throught the already existing project.

DvdHalliday committed:
Part 1

Added reference to the constructor and the getters return type, I also added here the copy constructor and the empty destructor

DvdHalliday committed:
Part 1 complement

Added the same stuff to the FDate class which is mainly used in the FVideogame class

DvdHalliday committed:
Part 2

Modified the category class files in order to make them reference-friendly. I also modified the GetGame into GetGameAt as well as following indications on what its return type should be and added an out parameter.

DvdHalliday committed:
Part 2  complement

Fixed the use of the GetGameAt occurrences in the steam.cpp file.

DvdHalliday committed:
Part 3

Added the same modifications for the classes CategoryContainer and Steam.
Changed how some functions work, especially the GetValidatedInput are now ValidateInput and don't return anything, they instead make use of an out parameter.

Sergio6040 commented on c++ training/SteamProject/Videogame/Category.cpp:
```diff
@@ -5,11 +5,16 @@ FCategory::FCategory() : Name("Default Name"), CurrentNumberOfGames(0)
-void FCategory::AddGame(const FVideogame Videogame)
+FCategory::FCategory(const FCategory& Other) : Name(Other.Name), CurrentNumberOfGames(Other.CurrentNumberOfGames)
+{
+	memcpy_s(Games, sizeof Games, Other.Games, sizeof Other.Games);
```
Great way to copy the array 👌

Sergio6040 commented on c++ training/SteamProject/Videogame/Category.h:
```diff
@@ -11,12 +11,15 @@ class FCategory
+	~FCategory() {}
+
+	void AddGame(const FVideogame& Videogame);
 	bool IsFull() const;
```
I think the return type could be const
```
const bool IsFull() const;
```

DvdHalliday replied:
Nope, the coding standards state that we should "Never use const on a return type" other than in the reference or pointers' return types.

Sergio6040 commented on c++ training/SteamProject/Videogame/Steam.h:
```diff
@@ -3,7 +3,7 @@
 
 
-class FSteam 
+class FSteam
```
this class is gigantic! may be is time to refactor the code :)

Sergio6040 commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -8,7 +8,10 @@ void FSteam::OpenMainMenu()
 
-	switch ((EMainMenuCommand)GetValidatedInput(MainMenuMessage, 1, 4))
+	int Choice;
+	ValidateInput(Choice, MainMenuMessage, 1, 4);
```
Great change! Easier to read

JDAlzate replied:
I agree that it's easier to read, but rather than it being about using an out parameter, it's about giving the value a name.

So for the previous case, it may had been easier to read to do something like
```
const int Choice = GetValidatedInput(/*blah blah*/);
switch ((EMainMenuCommand)Choice)
```
or even
```
const EMainMenuCommand Choice = (EMainMenuCommand)GetValidatedInput(/** blah blah */);
switch (Choice)
```

Sergio6040 commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -260,63 +291,62 @@ bool FSteam::AskChooseCategory() const
 
-	std::string UserInput;
-	std::cin >> UserInput;
+	std::cin >> OutInput;
```
what if the input has spaces?

Sergio6040 commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -260,63 +291,62 @@ bool FSteam::AskChooseCategory() const
+const int& FSteam::GetValidDay(const int& Month, const int& Year) const
 {
+	int Day;
 	switch (Month)
```
the empty cases can be excluded

DvdHalliday replied:
I don't think so, no. They are being used as shown in a lecture, that way january and march are being treated alike.

Sergio6040 reviewed (commented):
Great work in general, made some comments, I think is time to refactor some code 😎

DvdHalliday committed:
Sergio's code review commit

Changed the string ValidateInput so the program is able to take in the whole line instead of merely the first word before a space, i also had to add some extra clears and ignores in the cin 'cause i was now getting some kind of error when i didn't use them

JDAlzate commented on c++ training/SteamProject/Videogame/Date.h:
```diff
@@ -10,13 +10,16 @@ class FDate
 public:
 	FDate();
-	FDate(const int NewDay, const int NewMonth, const int NewYear);
+	FDate(const int& NewDay, const int& NewMonth, const int& NewYear);
```
References are useful to avoid potentially hard-to-copy types. This will never be the case with built-in types, so for built-in types, it's better to pass them by value.

JDAlzate commented on c++ training/SteamProject/Videogame/Date.h:
```diff
@@ -10,13 +10,16 @@ class FDate
-	int GetYear() const;
+	void SetDate(const int& Day, const int& Month, const int& Year);
+
+	const int& GetDay() const;
```
The above comment is also true for these return types. There never is a need to use references to built-in types if your only purpose is to avoid a copy.

JDAlzate commented on c++ training/SteamProject/Videogame/Videogame.h:
```diff
@@ -11,11 +11,14 @@ class FVideogame
 	std::string GetFormattedDate() const;
-	std::string GetName() const;
-	std::string GetStudio() const;
+	const std::string& GetName() const;
```
Now this is a good use of references as a return type since strings are structures that are expensive to copy.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.h:
```diff
@@ -20,31 +20,34 @@ class FSteam
-	std::string GetValidatedInput(const std::string Message) const;
+	bool IsInteger(const float& Float) const;
+	bool IsInRange(const float& Value, const int& LowerBound, const int& UpperBound) const;
+	void ValidateInput(int& OutInput, const std::string& Message, const int& LowerBound, const int& UpperBound) const;
```
Ouput parameters are useful in some scenarios, like for example when you want to return a value but also return whether or not the returned value is valid, so you use a bool return type and an out parameter for the other type you want to return.

However, they are not the standard way of returning stuff, that's what the return type is for. With this in mind, although this method works fine, I believe that it would be clearer if you used the return type, the way you were doing before.

Ouput parameters should be special cases, rather than the rule. Especially with built-in types.

JDAlzate reviewed (changes requested):
Very nice :) Couple comments there

DvdHalliday committed:
JD's code review commit

Took off all reference in built-in parameters and return types. Changed the ValidateInput functions back to how they used to work as well as their behaviours on the FSteam methods.
